From nothing to something

davidguhao opened the pull request:
The main branch only contains the basic framework provided by Android Studio. Now I have finished: 

* Main screen with the list of upcoming movies
* Create a details screen with the following movie data: name, poster image, list of genres, overview, and release date.


Should there be some other requirements that I haven't done or haven't noticed. Just let me know.

Commit:
Basic build...

On branch development
Changes to be committed:
	new file:   .idea/compiler.xml
	modified:   .idea/gradle.xml
	new file:   .idea/kotlinc.xml
	modified:   .idea/misc.xml
	modified:   app/build.gradle.kts
	modified:   app/src/main/java/com/example/squareonetestproject/MainActivity.kt
	new file:   app/src/main/java/com/example/squareonetestproject/MovieRepository.kt
	new file:   app/src/main/java/com/example/squareonetestproject/UseCase.kt

Commit:
Finished the basic functions.

Commit:
Getting it more beautiful.

guilucas07 commented on app/src/main/java/com/example/squareonetestproject/MainActivity.kt:
```diff
@@ -22,25 +58,98 @@ class MainActivity : ComponentActivity() {
     }
+}
+
+class MainViewModel(useCase: UseCase) : ViewModel() {
+    class Factory(private val useCase: UseCase): ViewModelProvider.Factory {
+        override fun <T : ViewModel> create(modelClass: Class<T>): T {
+            return MainViewModel(useCase) as T
+        }
+    }
+
+    val movies = useCase.getMovies().stateIn(
+        viewModelScope, SharingStarted.Lazily, listOf())
```
Any special reason to do it manually instead of using a dependency injection tool?

davidguhao replied:
No special reason. Currently it is using Hilt

guilucas07 commented on app/src/main/java/com/example/squareonetestproject/MovieRepository.kt:
```diff
@@ -0,0 +1,93 @@
+                            Request.Builder()
+                                .url("https://api.themoviedb.org/3/discover/movie?language=en-US&page=1&sort_by=primary_release_date.desc")
+                                .get()
+                                .addHeader("accept", "application/json")
+                                .addHeader("Authorization", "Bearer $authToken")
+                                .build()
+                        ).execute().body!!.string(),
+                        ResponseFromNetwork::class.java
```
Any special reason to do it manually instead of using a tool such as Retrofit?

davidguhao replied:
No special reason, just use any kind of method that sound direct to me. And currently it is using Retrofit already

guilucas07 commented on app/src/main/java/com/example/squareonetestproject/MovieRepository.kt:
```diff
@@ -0,0 +1,93 @@
+
+            }
+
+            delay(5000)
```
what is the reason for that?

davidguhao replied:
Removed

guilucas07 commented on app/src/main/java/com/example/squareonetestproject/MovieRepository.kt:
```diff
@@ -0,0 +1,93 @@
+        val authToken = "eyJhbGciOiJIUzI1NiJ9.eyJhdWQiOiI3YmQ2ZDZjNmQ2ZDIwZWQzMjgyZWQ4ODkzZmM0Y2JjNCIsInN1YiI6IjY1NDIwZjk5YTU4OTAyMDBhZDNlYmU3NyIsInNjb3BlcyI6WyJhcGlfcmVhZCJdLCJ2ZXJzaW9uIjoxfQ.3Ajiics-oMoYqnzh61l-VLbchzfUrPXoXkRTYIhrD08"
+
+
+        while(true) {
```
why is this needed?

davidguhao replied:
Removed

guilucas07 reviewed (commented):
Hi David, thanks for taking the time to do this project.

I asked some questions and have some considerations/requests:

1. You're not using the correct endpoint. I asked to get upcoming movies. Give a second look at TMDB documentation
2. You are showing the genres ids on the movies details. Isn't possible to get the names?
3. movie posters have different sizes in the app. Isn't it to possible make them, at least, have a similar size?
4. you are fetching just the first page of the list. Isn't possible to load the others pages? Could you implement an endless list of upcoming movies?
5. Could you create a separate screen for movie details and show the backdrop image when it's available (instead of using the poster image)?

Commit:
Basically finished writing all related features [without testing]

Commit:
Testing finished, bug fixed.

1 Used the correct endpoint right now.
2 Translated the id to their names.
3 Fix the movie posters size.
4 Let it become an endless list
5 Create another activity to as a separate screen for movie details and show the backdrop image instead of the poster image.
6 Used dependency injection tool - Hilt
7 Used Retrofit

davidguhao commented:
Hi Gui,

Thank you for inspecting my code carefully and I pushed another 2 commits according your comments. Here's a summary:

    1 Used the correct endpoint right now.
    2 Translated the id to their names.
    3 Fix the movie posters size.
    4 Let it become an endless list
    5 Create another activity to as a separate screen for movie details and show the backdrop image instead of the poster image.
    6 Used dependency injection tool - Hilt
    7 Used Retrofit

Should there be some other problems I omit, I am more than happy to fix them.

guilucas07 commented on app/src/main/java/com/example/squareonetestproject/MovieRepository.kt:
```diff
@@ -0,0 +1,117 @@
+                // Get genre map from network
+                try {
+                    CoroutineScope(Dispatchers.IO).async {
+                        tmdb.getGenreMap()
```
change it to 
`tmdb.getGenreMap().also{
Log.d("movieRepository", "request genre map")
}` and check `Logcat` tab on Android Studio

guilucas07 replied:
Take a look on `HttpLoggingInterceptor`. It can be useful too

davidguhao replied:
Using HttpLoggingInterceptor, it's pretty nice and neat method to log http request!<3

guilucas07 commented on app/src/main/java/com/example/squareonetestproject/MovieRepository.kt:
```diff
@@ -0,0 +1,117 @@
+            // Get upcoming movies from network, first we need to get some basic info.
+            val upComingMovies = try {
+                CoroutineScope(Dispatchers.IO).async {
+                    tmdb.getUpcoming(1)
```
similar here
`tmdb.getUpcoming(1).also {Log.d("movieRepository", "request upcoming page 1")}`

guilucas07 commented on app/src/main/java/com/example/squareonetestproject/MovieRepository.kt:
```diff
@@ -0,0 +1,117 @@
+                    result.addAll(
+                        try {
+                            CoroutineScope(Dispatchers.IO).async {
+                                tmdb.getUpcoming(currentPage)
```
similar here
`tmdb.getUpcoming(currentPage).also {
                                    Log.d("movieRepository", "request upcoming page $currentPage")}`

Commit:
1 Using HttpLoggingInterceptor
2 Delay the network request speed since it doesn't need to catch up with that speed

davidguhao commented:
Since you mentioned HttpLoggingInterceptor, I embedded this tool to log network request. It's pretty nice and neat to log everything by this way since all network requests are using Retrofit and Retrofit is using OkHttp at the same time. 

I pushed another one commit to make it better! 

1 Using HttpLoggingInterceptor
2 Delay the network request speed since it doesn't need to catch up with that speed


Cheers!!!!!
